OAuth/ResourceOwner/GenericOAuth2ResourceOwner.php

```diff
@@ -59,18 +60,14 @@ public function getUserInformation(array $accessToken, array $extraParameters =
     public function getAuthorizationUrl($redirectUri, array $extraParameters = [])
     {
         if ($this->options['csrf']) {
-            if (null === $this->state) {
-                $this->state = $this->generateNonce();
-            }
-
-            $this->storage->save($this, $this->state, 'csrf_state');
+            $this->handleCsrfToken();
         }
 
         $parameters = array_merge([
             'response_type' => 'code',
             'client_id' => $this->options['client_id'],
             'scope' => $this->options['scope'],
-            'state' => $this->state ? urlencode($this->state) : null,
+            'state' => $this->state->encode() ?: null,
```

Comment:
reading code below it will never reach the `null` cause encode always return something.

Reply:
@stloyd This actually makes sense as is, because it's a shorthand ternary operator and not a null coalescing operator.
```php
// results in null if encode is an empty string meaning the state will not be passed in the URL
'state' => $this->state->encode() ?: null
```

```php
// results in always having a state parameter in the URL, even if it's empty if CSRF is disabled
'state' => $this->state->encode()
```

We should not pass empty `state` parameters to the oAuth2 provider, so I'm voting to keep the shorthand ternary op.

Reply:
Problem is that implemetation of `encode` never returns empty string, if it returns such, then there should be validation & return `null` in such case not here.

Reply:
This means that the `encode` method should be allowed to return `null`. Is that correct?

Reply:
IMHO yes, otherwise I see some design flaw, cause if we say it may return empty string but as such is not acceptable & we replace it with `null`, then from design it should never return empty string nor convert empty array with base64+json, but in such case return simple `null`.

Reply:
@stloyd Done, please review. I've added 2 tests also for instantiating the state with empty values, https://github.com/balazscsaba2006/HWIOAuthBundle/blob/d21706afd566146023558511a1d1bd6f5735e1db/Tests/OAuth/State/StateTest.php#L106-L113

Reply:
@stloyd Any feedback?